external/llvm-project/mlir/include/mlir/ExecutionEngine/SystemDevices.h

```diff
@@ -0,0 +1,75 @@
+//===- SystemDevices.h - ExecutionEngine System Devices --------*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides a system device interface for the execution engine.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_EXECUTIONENGINE_SYSTEMDEVICES_H_
+#define MLIR_EXECUTIONENGINE_SYSTEMDEVICES_H_
+
+#include <string>
+#include <unordered_map>
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringMap.h"
+
+namespace mlir {
+
+/// SystemDevice captures device specifics.
+///
+
+struct SystemDevice {
+  enum class Type : int32_t { ECPU, EGPU, ENPU, EALT };
+
+  Type type;
+  llvm::SmallString<32> chip;
+  uint32_t count = 1;
+  llvm::StringMap<llvm::SmallString<8>> properties = {};
+};
+
+/// SystemDevices captures devices for the current system.
+///
+class SystemDevices : public llvm::StringMap<const SystemDevice &> {
```

Comment:
If we're going to have a map of pointers, how do we ensure lifetimes? That is, can we get a better sense of who owns this data and how long it'll live for?

Secondly, is `SystemDevices` really a map? It feels better for it to have a map and to provide wrappers around the map API that fit our usecase.